### ISSUE TITLE
Add in hooks and filters for numerous plugin operations to facilitate supporting WordPress installations using Amazon S3 and other cloud storage

### ISSUE THREAD

**Issue by Vheissu** [Shorthand/shorthand_connect_wordpress#15]
On our site, we make use of AWS for our hosting. In our situation, we have epithermal storage and, thus, no fixed storage. This poses a problem when using the copying functionality that copies over the assets from the Shorthand API and unzips them. At present, the files will get moved into the temporary storage inside the `wp-content/shorthand` folder. Still, when an instance is killed, or a new instance is created as part of our autoscaling configuration, those assets will no longer be available, and the site breaks.

I do know that Shorthand offers S3 integration itself on the platform. However, to my knowledge, this doesn't extend to the Shorthand Connect plugin. From what I can see, it makes an API request for a `.zip` file, unpacks it locally, saves the path, and then does some `file_get_contents` calls to get the content.

I will have a pull request coming shortly, which will be quite basic. Essentially, I will be adding some actions and filters into the plugin, which will allow me to remap URLs to Amazon S3 and the ability to get the downloaded files from the API and then copy them over to an Amazon S3 bucket.

Before I contribute these back as a pull request, I'll be testing them on a real site, using Shorthand Connect to ensure that nothing breaks.

**Comment by benhoad** [comment]
apologies for the delay - that's now released in 1.3.21

We do offer s3 pushing for Drupal, where it's possible to split assets out to separate API-Only publishing config. Though Wordpress is yet get this feature as it's only been an issue thus far with VIP customers.

Out of curiosity what do you do currently for the AWS remapping?

**Comment by Vheissu** [comment]
@benhoad Love it. Thank you.

For the AWS remapping, this is what I am currently doing. Basically, changing the URL to map from `sitedomain.com/wp-content` to an S3 URL structure that replicates the `wp-content` path on S3. And then I make a call to get the HTML files from S3 with some other filters.

```
// Fix up path to map to S3 structure
add_filter('sh_get_story_path', function($path, $destination) {
    $upload_dir = wp_upload_dir();

    $path = str_ireplace($upload_dir['path'], '', $path);
    $path = 'https://contentbucket-bucket.s3.ap-southeast-2.amazonaws.com' . $path;

    return $path;
}, 10, 2);

// Rewrite from current site content URL to S3 content URL
add_filter('sh_get_story_url', function($url) {
    $path = str_ireplace('https://sitename.com/wp-content/', 'https://contentbucket-bucket.s3.ap-southeast-2.amazonaws.com/shorthand/', $url);

    return $path;
}, 10, 2);

// Load article file from S3 using file_get_contents
add_filter('sh_pre_process_body', function($body, $assets_path, $article_file) {
    $load_file = file_get_contents($assets_path . '/article.html');

    return shand_fix_content_paths($assets_path, $load_file);
}, 10, 3);

// Load head file from S3 using file_get_contents
add_filter('sh_pre_process_head', function($head, $assets_path, $head_file) {
    $load_file = file_get_contents($assets_path . '/head.html');

    return shand_fix_content_paths($assets_path, $load_file);
}, 10, 3);
```
